linera-core/src/chain_worker/state/attempted_changes.rs

```diff
@@ -131,8 +131,17 @@ where
         local_time: Timestamp,
     ) -> Result<(), WorkerError> {
         // Create the vote and store it in the chain state.
+        let executed_block = outcome.with(proposal.content.block.clone());
+        let blobs = if proposal.validated_block_certificate.is_some() {
+            self.state
```

Comment:
Why if the proposal has a `ValidatedBlockCertificate` on it we extract blobs and otherwise not? It still has `blobs` in it (like in `proposal.blobs`.

Reply:
Yes, but we only set the locked block in the chain manager if the proposal has a validated block certificate. (That validated block from an earlier round becomes the locked block.) So we only need its blobs in that case.

The proposal itself is also stored in the chain manager itself, but in a different field. For it, we currently only store the published blobs, because it's not critical for liveness. (Or rather: it only is for `Fast` blocks, for which we enforce that the published blobs _are_ all the required blobs.)

Reply:
We do the same check `if let Some(lite_cert) = &proposal.validated_block_certificate {` inside the `create_vote` already - can't you extract the "locked blobs" there instead?

Reply:
No, because we may have to read them from storage. I'd really like to avoid passing a storage handle into `create_vote`.

Reply:
I'd say that's still better than current implementation which requires a lot of context which isn't in the code or in the comments.

Reply:
So you'd clone `self.state.storage` here and pass it into `create_vote`, rather than calling `get_required_blobs` here?

Then I should probably change `get_required_blobs` to _not_ take a `ChainWorkerState` receiver; instead, it would take a `&ChainManager` and a `&Storage`. :thinking:

Reply:
Maybe it should be a method of `ChainManager` then?

Reply:
Oh, no, I can't. The `ChainManager` is in `linera-chain` which doesn't even depend on storage! (And storage depends on _it_.)

Reply:
Oh, that's weird. I'd say dependencies should be the other way around (chain depends on storage).

Reply:
Yes, that's another discussion to have with the team some time.

Reply:
OK, another idea: I move the whole `if let Some(lite_cert)` block out of `create_vote`, to the call site, i.e. we put the locked block and its blobs together there. Then we call a new `ChainManager::set_locked` method explicitly before calling `create_vote`?

Reply:
That doesn't look great either: https://github.com/afck/linera-protocol/commit/c0fda9906f9c1b8c0de762aaa46ae1f5faad8220

Reply:
I agree - if we can't pass the `Storage` in any way there it will be ugly 😬